src/index/client/oauth2.ts

```diff
@@ -0,0 +1,678 @@
+
+	/**
+	 * The refresh token.
+	 */
+	refreshToken: string;
+
+}
+
+export interface RevocationOptions {
+
+	/**
+	 * If true, uses a GET rather than POST request to revoke an access token.
+	 */
+	useGet?: boolean;
+
+}
+
+export interface TokenGrantorParams extends HasClientId {
+
+	/**
+	 * The client secret for your application.
+	 */
+	clientSecret?: string;
+
+}
+
+export type GrantParams = AuthCodeGrantParams | JwtGrantParams | RefreshGrantParams;
+
+const DEFAULT_GRANT_OPTIONS = Object.freeze({
+	decodeIdToken: true,
+	parseUserInfo: true,
+	responseFormat: 'application/json',
+	verifySignature: true
+});
+
+const DEFAULT_REVOCATION_OPTIONS = Object.freeze({
+	useGet: false
+});
+
+const DEFAULT_REQUEST_CONFIG = Object.freeze({
+	validateStatus: undefined
+});
+
+/**
+ *  An OAuth 2.0 client that implements the [OAuth 2.0 Authorization Framework](https://tools.ietf.org/html/rfc6749) specification.
+ */
+export class OAuth2Client implements AuthClient {
+
+	/**
+	 * The client name.
+	 */
+	protected clientType: string;
+
+	/**
+	 * [Authorization Endpoint](https://tools.ietf.org/html/rfc6749#section-3.1)
+	 */
+	protected authorizationEndpoint?: string;
+
+	/**
+	 * [Token Endpoint](https://tools.ietf.org/html/rfc6749#section-3.2)
+	 */
+	protected tokenEndpoint?: string;
+
+	/**
+	 * [Revocation Endpoint](https://tools.ietf.org/html/rfc7009#section-2)
+	 */
+	protected revocationEndpoint?: string;
+
+	/**
+	 * The environment parameters used for this client.
+	 */
+	protected readonly env: Readonly<Environment>;
+
+	/**
+	 * Creates a new OAuth2 client with the given environment.
+	 *
+	 * @param env The OAuth2 environment parameters.
+	 */
+	constructor(env: Environment) {
+		this.clientType = 'OAuth2';
+		this.env = Object.freeze(Object.assign({}, env));
+	}
+
+	/**
+	 * @inheritdoc
+	 */
+	public async init() {
+
+		if (!this.env.authorizationEndpoint) {
+			throw new Error('Missing required string parameter: env[authorizationEndpoint]');
+		}
+
+		if (!this.env.revocationEndpoint) {
+			throw new Error('Missing required string parameter: env[revocationEndpoint]');
+		}
+
+		if (!this.env.tokenEndpoint) {
+			throw new Error('Missing required string parameter: env[tokenEndpoint]');
+		}
+
+		this.authorizationEndpoint = this.env.authorizationEndpoint;
+		this.revocationEndpoint = this.env.revocationEndpoint;
+		this.tokenEndpoint = this.env.tokenEndpoint;
+
+	}
+
+	/**
+	 * @inheritdoc
+	 */
+	public createAuthorizationUrl(params: AuthUrlParams, opts?: AuthOptions) {
+
+		if (!this.authorizationEndpoint) {
+			throw new Error(`${this.clientType} client has not been initialized`);
+		}
+
+		// To avoid any inadvertent parameter leaks, create a new object from which to generate the request.
+		const internalOpts: AuthOptions = {
+			display: opts && opts.display,
+			immediate: opts && opts.immediate,
+			prompt: opts && opts.prompt,
+			state: opts && opts.state
+		};
```

Comment:
You can use spread operator to make a shallow copy like this:

```
const internalOpts: AuthOptions = { ...opts };
```
This is the same as doing `Object.assign({}, opts);`

You might need to do the undefined check also, but at least it won't need to do it for each prop. The values are all primitives so it should be okay, but be careful if we ever add an object.

Reply:
Extra clarification: This was because what is currently happening here is a shallow copy of everything on that particular interface. What I've suggested would copy all props which you may not want if you're trying to pick and choose.

Reply:
Indeed, I only want to copy these options.